addons/payment_asiapay/models/payment_transaction.py

```diff
@@ -0,0 +1,139 @@
+
+        rendering_values.update({
+            "api_url": self.acquirer_id._asiapay_get_api_url()
+        })
+        return rendering_values
+
+    @api.model
+    def _get_tx_from_feedback_data(self, provider, data):
+        tx = super()._get_tx_from_feedback_data(provider, data)
+        if provider != "asiapay":
+            return tx
+
+        reference = data.get('Ref')
+        if not reference:
+            raise ValidationError(
+                "AsiaPay: " + _(
+                    "Received data with missing reference %(r)s.",
+                    r=reference
+                )
+            )
+
+        order_reference = reference.split()[0]
+        tx = self.search([('reference', '=', order_reference), ('provider', '=', 'asiapay')])
+        if not tx:
+            raise ValidationError(
+                "AsiaPay: " + _("No transaction found matching reference %s.", order_reference)
+            )
+
+        return tx
+
+    def _process_feedback_data(self, data):
+        super()._process_feedback_data(data)
+        if self.provider != "asiapay":
+            return
+        if float_compare(float(data.get('Amt', '0.0')), self.currency_id._convert(self.amount, self.acquirer_id.asiapay_currency_id, self.company_id, fields.Date.today()), 2) != 0:
+            logging_values = {
+                'amount': data.get('Amt', '0.0'),
+                'total': self.currency_id._convert(self.amount, self.acquirer_id.asiapay_currency_id, self.company_id, fields.Date.today()),
+                'reference': self.reference,
+            }
+            _logger.error(
+                "the paid amount (%(amount)s) does not match the total + fees (%(total)s + "
+                "%(fees)s) for the transaction with reference %(reference)s", logging_values
+            )
+            raise ValidationError("AsiaPay: " + _("The amount does not match the total + fees."))
+        if secure_hash := data.get('secureHash'):
+            args = {
+                'hash_function': self.acquirer_id.asiapay_secure_hash_function,
+                'secret_hash': secure_hash,
+                'src': data.get('src'),
+                'prc': data.get('prc'),
+                'success_code': data.get('successcode'),
+                'merchant_reference': data.get('Ref'),
+                'paydollar_reference': data.get('PayRef'),
+                'curr_code': data.get('Cur'),
+                'amount': data.get('Amt'),
+                'payer_authentication_status': data.get('payerAuth'),
+                'secret': self.acquirer_id.asiapay_secure_hash,
+            }
+            if not payment_asiapay_utils.verify_date_feed(**args):
+                raise ValidationError(
+                    "Alipay: " + _("The secure hash does not match the configured AsiaPay account.")
+                )
+
+        status = data.get('successcode')
+        if status == '0':
+            self._set_done()
+        elif status == '1':
+            self._set_canceled()
```

Comment:
Is this the exhaustive list of success codes that the provider can send?

Reply:
![image](https://user-images.githubusercontent.com/85471608/185552580-0b0bb7c0-711e-49de-a475-ad6bb2b92c82.png)

They only have 3 status